From ab26e16aeca361c30a8098b8c886a81d10f36ba6 Mon Sep 17 00:00:00 2001
From: Vinaykumar Masane <vinaykumar.masane@nxp.com>
Date: Mon, 1 Feb 2016 12:56:27 +0530
Subject: [PATCH 32/60] SBR-705 Added Barebox support to Parse Reset Key Press

If Reset Key is Pressed for min 3 sec, halt at Barebox Prompt
 with DIAG LED SOLID RED
Else continue to boot further

Signed-off-by: Vinaykumar Masane <vinaykumar.masane@nxp.com>
---
 arch/arm/boards/comcerto-rv340/env/bin/boot        |  3 +-
 arch/arm/boards/comcerto-rv340/env/bin/init        | 25 +++++++-
 arch/arm/boards/comcerto-rv340/env/config          | 10 ++++
 arch/arm/configs/comcerto-2k_rv340_defconfig       |  4 ++
 arch/arm/configs/comcerto-2k_rv340_diags_defconfig |  4 ++
 .../configs/comcerto-2k_rv340_uloader_defconfig    |  1 +
 arch/arm/mach-comcerto/Kconfig                     |  2 +
 commands/Kconfig                                   |  8 +++
 commands/Makefile                                  |  1 +
 commands/check_reset.c                             | 67 ++++++++++++++++++++++
 10 files changed, 121 insertions(+), 4 deletions(-)
 create mode 100644 commands/check_reset.c

diff --git a/arch/arm/boards/comcerto-rv340/env/bin/boot b/arch/arm/boards/comcerto-rv340/env/bin/boot
index 091e87e..9d5334e 100755
--- a/arch/arm/boards/comcerto-rv340/env/bin/boot
+++ b/arch/arm/boards/comcerto-rv340/env/bin/boot
@@ -2,8 +2,7 @@
 
 . /env/config
 
-ROOTFS_PART1=4
-ROOTFS_PART2=7
+gpio_set_value $GPIO_LED_DIAG $OFF 
 
 echo "Setting env LAN, WAN1, WAN2..."
 boardinfo -m
diff --git a/arch/arm/boards/comcerto-rv340/env/bin/init b/arch/arm/boards/comcerto-rv340/env/bin/init
index 7f42f8c..ea0e840 100755
--- a/arch/arm/boards/comcerto-rv340/env/bin/init
+++ b/arch/arm/boards/comcerto-rv340/env/bin/init
@@ -36,12 +36,33 @@ if [ -f /env/bin/init_board ]; then
 	/env/bin/init_board
 fi
 
+# Parse Reset Key Press
+chkreset
+if [ $? -eq 0 ]; then
+  val_gpio_reset="$val_gpio_reset"
+  if [ "X$val_gpio_reset" = "Xclear" ]; then
+    timeout -a $autoboot_timeout
+    chkreset
+    if [ $? -eq 0 ]; then
+      val_gpio_reset="$val_gpio_reset"
+      if [ "X$val_gpio_reset" = "Xclear" ]; then
+        gpio_direction_output $GPIO_LED_DIAG $ON
+        gpio_set_value $GPIO_LED_DIAG $ON
+        . /env/bin/_update_help
+        exit
+      fi
+    fi
+  fi
+fi
+
 echo
 echo -n "Hit any key to stop autoboot: "
 timeout -a $autoboot_timeout
 if [ $? != 0 ]; then
-	. /env/bin/_update_help
-	exit
+  gpio_direction_output $GPIO_LED_DIAG $ON
+  gpio_set_value $GPIO_LED_DIAG $ON
+  . /env/bin/_update_help
+  exit
 fi
 
 boot
diff --git a/arch/arm/boards/comcerto-rv340/env/config b/arch/arm/boards/comcerto-rv340/env/config
index f95805b..4276658 100644
--- a/arch/arm/boards/comcerto-rv340/env/config
+++ b/arch/arm/boards/comcerto-rv340/env/config
@@ -37,9 +37,19 @@ imageavc=root_avcsign.jffs2
 imageweb=root_webrootdb.jffs2
 imagelic=root_license.jffs2
 
+# active inactive image identifiers
 active=image1
 inactive=image2
 
+# Rootfs Partition Identifiers
+ROOTFS_PART1=4
+ROOTFS_PART2=7
+
+# GPIO Constants
+GPIO_LED_DIAG=50
+ON=0
+OFF=1
+
 fast_spi_device="spi1.0"
 fast_spi_parts="128k(uloader)ro,512k(barebox),128k(env),128k(boardinfo)ro"
 
diff --git a/arch/arm/configs/comcerto-2k_rv340_defconfig b/arch/arm/configs/comcerto-2k_rv340_defconfig
index c3641e7..2ffc691 100644
--- a/arch/arm/configs/comcerto-2k_rv340_defconfig
+++ b/arch/arm/configs/comcerto-2k_rv340_defconfig
@@ -53,6 +53,7 @@ CONFIG_COMCERTO_BOOTLOADER=y
 CONFIG_COMCERTO_NAND=y
 CONFIG_COMCERTO_SERDES=y
 CONFIG_COMCERTO_I2C=y
+CONFIG_COMCERTO_GPIO=y
 CONFIG_COMCERTO_SATA=y
 CONFIG_COMCERTO_SECUREBOOT=y
 # CONFIG_SECUREBOOT_TEST is not set
@@ -103,6 +104,7 @@ CONFIG_HAS_KALLSYMS=y
 CONFIG_HAS_MODULES=y
 CONFIG_CMD_MEMORY=y
 CONFIG_ENV_HANDLING=y
+CONFIG_GENERIC_GPIO=y
 
 #
 # General Settings              
@@ -249,9 +251,11 @@ CONFIG_CMD_TEST=y
 CONFIG_CMD_VERSION=y
 CONFIG_CMD_HELP=y
 CONFIG_CMD_DEVINFO=y
+CONFIG_CMD_GPIO=y
 CONFIG_CMD_UNLZO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_USB=y
+CONFIG_CMD_CHKRESET=y
 
 #
 # c2000			
diff --git a/arch/arm/configs/comcerto-2k_rv340_diags_defconfig b/arch/arm/configs/comcerto-2k_rv340_diags_defconfig
index 23b7498..59c1d3d 100644
--- a/arch/arm/configs/comcerto-2k_rv340_diags_defconfig
+++ b/arch/arm/configs/comcerto-2k_rv340_diags_defconfig
@@ -53,6 +53,7 @@ CONFIG_COMCERTO_DIAG=y
 CONFIG_COMCERTO_NAND=y
 CONFIG_COMCERTO_SERDES=y
 CONFIG_COMCERTO_I2C=y
+CONFIG_COMCERTO_GPIO=y
 CONFIG_COMCERTO_SATA=y
 CONFIG_COMCERTO_SECUREBOOT=y
 # CONFIG_SECUREBOOT_TEST is not set
@@ -141,6 +142,7 @@ CONFIG_HAS_KALLSYMS=y
 CONFIG_HAS_MODULES=y
 CONFIG_CMD_MEMORY=y
 CONFIG_ENV_HANDLING=y
+CONFIG_GENERIC_GPIO=y
 
 #
 # General Settings              
@@ -289,9 +291,11 @@ CONFIG_CMD_TEST=y
 CONFIG_CMD_VERSION=y
 CONFIG_CMD_HELP=y
 CONFIG_CMD_DEVINFO=y
+CONFIG_CMD_GPIO=y
 CONFIG_CMD_UNLZO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_USB=y
+CONFIG_CMD_CHKRESET=y
 
 #
 # c2000			
diff --git a/arch/arm/configs/comcerto-2k_rv340_uloader_defconfig b/arch/arm/configs/comcerto-2k_rv340_uloader_defconfig
index d7a2eef..f4e1ac8 100644
--- a/arch/arm/configs/comcerto-2k_rv340_uloader_defconfig
+++ b/arch/arm/configs/comcerto-2k_rv340_uloader_defconfig
@@ -231,6 +231,7 @@ CONFIG_CMD_BOOTB=y
 # CONFIG_CMD_HELP is not set
 # CONFIG_CMD_DEVINFO is not set
 # CONFIG_CMD_UNLZO is not set
+# CONFIG_CMD_CHKRESET is not set
 
 #
 # c2000			
diff --git a/arch/arm/mach-comcerto/Kconfig b/arch/arm/mach-comcerto/Kconfig
index f5c5d9e..3430073 100644
--- a/arch/arm/mach-comcerto/Kconfig
+++ b/arch/arm/mach-comcerto/Kconfig
@@ -38,6 +38,7 @@ config  MACH_COMCERTO_C2K_MFCNEVM
 
 config  MACH_COMCERTO_C2K_RV340
 	bool "RV340"
+  select GENERIC_GPIO
 	help
 
 config  MACH_COMCERTO_C2K_ASIC
@@ -87,6 +88,7 @@ config COMCERTO_BOOTLOADER
 	select COMCERTO_SERDES if !MACH_COMCERTO_C2K_RTSM
 	select COMCERTO_I2C
 	select COMCERTO_NAND if !MACH_COMCERTO_C2K_RTSM
+	select COMCERTO_GPIO if MACH_COMCERTO_C2K_RV340
 	select COMCERTO_SATA
 	select COMCERTO_SECUREBOOT
 	help
diff --git a/commands/Kconfig b/commands/Kconfig
index 6350ec3..12d5936 100644
--- a/commands/Kconfig
+++ b/commands/Kconfig
@@ -328,6 +328,14 @@ config CMD_RESET
 	tristate
 	prompt "reset"
 
+config CMD_CHKRESET
+	bool
+	depends on MACH_COMCERTO_C2K_RV340
+	prompt "RV340 Check Reset"
+	default n
+	help
+	  This command allows to update and show RV340 Board Specfic Information.
+
 config CMD_GO
 	tristate
 	prompt "go"
diff --git a/commands/Makefile b/commands/Makefile
index e87b641..88518d2 100755
--- a/commands/Makefile
+++ b/commands/Makefile
@@ -11,6 +11,7 @@ obj-$(CONFIG_CMD_EDIT)		+= edit.o
 obj-$(CONFIG_CMD_EXEC)		+= exec.o
 obj-$(CONFIG_CMD_SLEEP)		+= sleep.o
 obj-$(CONFIG_CMD_RESET)		+= reset.o
+obj-$(CONFIG_CMD_CHKRESET)		+= check_reset.o
 obj-$(CONFIG_CMD_GO)		+= go.o
 obj-$(CONFIG_NET)		+= net.o
 obj-$(CONFIG_CMD_PARTITION)	+= partition.o
diff --git a/commands/check_reset.c b/commands/check_reset.c
new file mode 100644
index 0000000..f3dafaa
--- /dev/null
+++ b/commands/check_reset.c
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2016 Vinaykumar Masane-B47602 <vinaykumar.masane@freescale>,
+ * Freescale Semicondunctors Pvt. Ltd.
+ *
+ * See file CREDITS for list of people who contributed to this
+ * project.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+/**
+ * @file
+ * @brief chkreset: During Boot-up Parse Reset key press for 3 sec
+ * @                If Reset Key found pressed (0) for min 3 sec,
+ * @                halt at barebox prompt with Diag LED Solid RED
+ * @                else continue to boot further
+ */
+
+#include <common.h>
+#include <command.h>
+#include <linux/ctype.h>
+#include <environment.h>
+#include <gpio.h>
+#include <mach/comcerto-2000.h>
+
+/**
+ * @param[in] cmdtp FIXME
+ * @param[in] argc Argument count from command line
+ * @param[in] argv List of input arguments
+ */
+static int do_chkreset(struct command *cmdtp, int argc, char *argv[])
+{
+  int value = 0;
+  const int gpio_reset = 4;
+
+  gpio_direction_input(gpio_reset);
+  value = gpio_get_value(gpio_reset);
+  if(value < 0)
+  {
+    printf("GPIO  failed returned = %d\n", value);
+    return COMMAND_ERROR;
+  }
+  setenv("val_gpio_reset", value == 0 ? "clear" : "set");
+  printf("GPIO # %d value = %d\n", gpio_reset, value);
+  return COMMAND_SUCCESS;
+}
+
+static const __maybe_unused char cmd_chkreset_help[] =
+"Usage: chkreset\n"
+;
+
+BAREBOX_CMD_START(chkreset)
+  .cmd    = do_chkreset,
+  .usage    = "check Reset Button Press Status",
+  BAREBOX_CMD_HELP(cmd_chkreset_help)
+BAREBOX_CMD_END
-- 
2.7.4

